### PR TITLE
Supports dynamic reading of ETH ABI information in external sdcrad an…

### PR DESCRIPTION
**Pull request by finalstatics**
Supports dynamic reading of ETH ABI information in external sdcrad and decode of transactions

### CODE REVIEW

File: app/src/main/java/com/keystone/cold/ui/fragment/main/EthTxConfirmFragment.java [hunk 1]
```diff
@@ -19,15 +19,19 @@
...
 
+import android.content.Context;
+import android.content.SharedPreferences;
 import android.graphics.Color;
 import android.os.Bundle;
 import android.os.Handler;
 import android.text.Spannable;
 import android.text.SpannableStringBuilder;
 import android.text.style.ForegroundColorSpan;
+import android.util.Log;
```
Comment (soralit) [inline app/src/main/java/com/keystone/cold/ui/fragment/main/EthTxConfirmFragment.java:30]: remove unnecessary import

File: app/src/main/java/com/keystone/cold/ui/fragment/main/EthTxConfirmFragment.java [hunk 2]
```diff
@@ -59,10 +62,12 @@
 import static com.keystone.cold.ui.fragment.setup.PreImportFragment.ACTION;
 
 public class EthTxConfirmFragment extends BaseFragment<EthTxConfirmBinding> {
-
+    private static final String VISITS_NAME = "visitsName";
+    private static final String SP_NAME = "EthTxConfirmFragment";
```
Comment (soralit) [inline app/src/main/java/com/keystone/cold/ui/fragment/main/EthTxConfirmFragment.java:66]: Manage preferences in https://github.com/KeystoneHQ/Keystone-cold-app/blob/main/app/src/main/java/com/keystone/cold/Utilities.java

Comment (soralit) [reply app/src/main/java/com/keystone/cold/ui/fragment/main/EthTxConfirmFragment.java:66]: Rename VISITS_NAME to a more business meaningful name.

File: app/src/main/java/com/keystone/cold/ui/fragment/main/EthTxConfirmFragment.java [hunk 3]
```diff
@@ -96,6 +103,27 @@ protected void init(View view) {
...
+        mBinding.ethTx.info.setOnClickListener(view1 -> realShowDialog());
+        showDialog();
+    }
+
+    private void showDialog() {
+        SharedPreferences sharedPreferences = mActivity.getSharedPreferences(SP_NAME, Context.MODE_PRIVATE);
+        int visits = sharedPreferences.getInt(VISITS_NAME, 0);
+        if (visits++ == 0) {
+            realShowDialog();
+            SharedPreferences.Editor editor = sharedPreferences.edit();
```
Comment (soralit) [inline app/src/main/java/com/keystone/cold/ui/fragment/main/EthTxConfirmFragment.java:115]: Manage preferences with `Utilities`

File: app/src/main/java/com/keystone/cold/ui/fragment/main/EthTxConfirmFragment.java [hunk 4]
```diff
@@ -221,18 +257,19 @@ private void updateAbiView(JSONObject abi) {
     protected void initData(Bundle savedInstanceState) {
 
     }
+
     public static SpannableStringBuilder highLight(String content) {
         SpannableStringBuilder spannable = new SpannableStringBuilder(content);
         Matcher matcher = pattern.matcher(spannable);
         while (matcher.find())
-            spannable.setSpan(new ForegroundColorSpan(0xff00cdc3), matcher.start() - 1 ,
+            spannable.setSpan(new ForegroundColorSpan(0xff00cdc3), matcher.start() - 1,
```
Comment (soralit) [inline app/src/main/java/com/keystone/cold/ui/fragment/main/EthTxConfirmFragment.java:265]: use r.color.icon_select

File: app/src/main/java/com/keystone/cold/ui/fragment/main/EthTxConfirmFragment.java [hunk 5]
```diff
@@ -221,18 +257,19 @@ private void updateAbiView(JSONObject abi) {
...
                     matcher.end() + 1, Spannable.SPAN_INCLUSIVE_INCLUSIVE);
 
         matcher = pattern1.matcher(spannable);
         while (matcher.find()) {
-            spannable.replace(matcher.start() - 1, matcher.start(),"(");
-            spannable.replace(matcher.end(), matcher.end() + 1,")");
-            spannable.setSpan(new ForegroundColorSpan(Color.RED), matcher.start() - 1 ,
+            spannable.replace(matcher.start() - 1, matcher.start(), "(");
+            spannable.replace(matcher.end(), matcher.end() + 1, ")");
+            spannable.setSpan(new ForegroundColorSpan(Color.RED), matcher.start() - 1,
```
Comment (soralit) [inline app/src/main/java/com/keystone/cold/ui/fragment/main/EthTxConfirmFragment.java:272]: r.color.red

File: app/src/main/res/layout/eth_tx_detail.xml [hunk 6]
```diff
@@ -276,37 +286,112 @@
             </LinearLayout>
 
             <LinearLayout
-                android:id="@+id/data"
+                android:id="@+id/undecode_data"
```
Comment (soralit) [inline app/src/main/res/layout/eth_tx_detail.xml:289]: undecoded_data

File: app/src/main/res/layout/eth_tx_detail.xml [hunk 7]
```diff
@@ -276,37 +286,112 @@
...
+                        android:layout_height="15dp"
+                        android:layout_marginStart="16dp"
+                        android:src="@drawable/direction" />
+
+                    <TextView
+                        android:layout_width="wrap_content"
+                        android:layout_height="wrap_content"
+                        android:layout_marginStart="3dp"
+                        android:layout_marginTop="3dp"
+                        android:text="@string/undecod_info"
```
Comment (soralit) [inline app/src/main/res/layout/eth_tx_detail.xml:332]: typo

File: app/src/main/res/layout/eth_tx_detail.xml [hunk 8]
```diff
@@ -276,37 +286,112 @@
...
+                        android:textColor="#FF0B0B"
```
Comment (soralit) [inline app/src/main/res/layout/eth_tx_detail.xml:333]: add new color in https://github.com/KeystoneHQ/Keystone-cold-app/blob/main/app/src/main/res/values/colors.xml

File: app/src/main/res/values-zh-rCN/strings.xml [hunk 9]
```diff
@@ -15,7 +15,7 @@
   ~ in the file COPYING.  If not, see <http://www.gnu.org/licenses/>.
   -->
 
-<resources>
+<resources xmlns:color="http://schemas.android.com/apk/res-auto">
```
Comment (soralit) [inline app/src/main/res/values-zh-rCN/strings.xml:18]: unnecessary code?

File: app/src/main/res/values-zh-rCN/strings.xml [hunk 10]
```diff
@@ -520,6 +520,11 @@
     <string name="calc_mnemonic_warning">您触发了“创建自己的助记词”这一高级功能，请确保您已充分了解助记词生成的相关知识后，再使用本功能。否则请关闭当前页面。</string>
     <string name="calc_mnemonic_warning_checkbox">我已充分了解随机熵、BIP39、私钥生成等相关知识。</string>
     <string name="note">须知</string>
+    <string name="tip">提示</string>
+    <string name="learn_more"><Data><![CDATA[Keystone支持解析交易，您可前往<font color="#00CDC3"><b> https://keyst.one/eth </b></font>了解详情。]]></Data></string>
+    <string name="undecod_info">注意！这是一个没有解密的数据。</string>
```
Comment (soralit) [inline app/src/main/res/values-zh-rCN/strings.xml:525]: typo

File: coinlib/src/main/java/com/keystone/coinlib/v8/ScriptLoader.java [hunk 11]
```diff
@@ -83,4 +93,70 @@ public static String readAsset(String fileName) {
         }
         return stringBuilder.toString();
     }
+
+    public static String getContentFromSdCard(String path, String fileName) {
+        if (TextUtils.isEmpty(externalSDCardPath())) {
+            Log.d(TAG, "sdCard is not exists");
```
Comment (soralit) [inline coinlib/src/main/java/com/keystone/coinlib/v8/ScriptLoader.java:99]: remove debug log. use info log if you want to log something in releases.

File: coinlib/src/main/java/com/keystone/coinlib/v8/ScriptLoader.java [hunk 12]
```diff
@@ -83,4 +93,70 @@ public static String readAsset(String fileName) {
...
+            }
+        }
+        return stringBuilder.toString();
+    }
+
+    private static String externalSDCardPath() {
+        String sdCardPath = "";
+        try {
+            StorageManager storageManager = (StorageManager) Coinlib.sInstance.getContext().getSystemService(Context.STORAGE_SERVICE);
+            // 7.0才有的方法
```
Comment (soralit) [inline coinlib/src/main/java/com/keystone/coinlib/v8/ScriptLoader.java:137]: comments in English

File: coinlib/src/main/java/com/keystone/coinlib/v8/ScriptLoader.java [hunk 13]
```diff
@@ -83,4 +93,70 @@ public static String readAsset(String fileName) {
...
+            Class<?> volumeClass = Class.forName("android.os.storage.StorageVolume");
+            Method getPath = volumeClass.getDeclaredMethod("getPath");
+            Method isRemovable = volumeClass.getDeclaredMethod("isRemovable");
+            getPath.setAccessible(true);
+            isRemovable.setAccessible(true);
+            StorageVolume storageVolume = storageVolumes.get(storageVolumes.size() - 1);
+            sdCardPath = (String) getPath.invoke(storageVolume);
+            Boolean isRemove = (Boolean) isRemovable.invoke(storageVolume);
+            if (storageVolumes.size() > 1) {
+                Log.d(TAG, "externalSDCardPath is === " + sdCardPath);
```
Comment (soralit) [inline coinlib/src/main/java/com/keystone/coinlib/v8/ScriptLoader.java:148]: remove debug logs.

**Review by soralit — CHANGES_REQUESTED** [review]

File: app/src/main/res/values-zh-rCN/strings.xml (hunk 10 above)
Comment (aaronisme) [reply app/src/main/res/values-zh-rCN/strings.xml:525]: this translation is not correct. not "解密" should be “解析”

**Review by soralit — APPROVED** [review]

**Pull request merged**